Multiple text formats for symbols via "format" expression (native port)

ChrisLoer opened the pull request:
Port of https://github.com/mapbox/mapbox-gl-js/pull/6994.

- [x] Port automatic argument coercion for compound expressions (this is no longer necessary for the `format` expression itself, but it was part of the JS PR and we want to keep native in sync)
- [x] Port `format` expression parsing code
- [x] Port format-aware shaping code
- [x] Implement Android conversion/serialization
- [x] Implement non-ICU styled Bidi for Qt
- [x] Figure out iOS and Android binding strategy: cc @fabian-guerra and @LukasPaczos. It shouldn't be too complicated, but like `collator`, `format` uses the new object/dictionary argument format e.g. `{ "font-scale": 1.5, "text-font": "awesome" }`. I think it should be just fine to defer binding work to later PRs.

ChrisLoer commented:
@asheemmamoowala if you want to start looking I just got an initial port of the expression code itself in...

1ec5 commented:
> Figure out iOS and Android binding strategy: … It shouldn't be too complicated, but like `collator`, `format` uses the new object/dictionary argument format e.g. `{ "font-scale": 1.5, "text-font": "awesome" }`.

The iOS/macOS runtime styling API will support this expression operator automatically via the ugly `MGL_FUNCTION()` fallback expression. #12739 tracks implementing a more convenient affordance of some sort in NSExpression.

Since this expression operator will be used to implement bilingual or multilingual labels, #12738 tracks enhancing the iOS/macOS expression localization functionality to avoid making such labels repeat the same name multiple times. (The same considerations apply to the Mapbox GL Language plugin and the Android map SDK’s localization plugin.)

/cc @nickidlugash

ChrisLoer commented:
Status update: the initial version of this is working 🎉 !

There's a lot of cleanup that remains to be done and it appears I'm occasionally doing some sort of undefined/unexpected behavior when I run the BiDi algorithm on styled text (note the random incorrect e -- I just sometimes see garbage like this show up):

<img width="284" alt="screenshot 2018-09-13 17 11 20" src="https://user-images.githubusercontent.com/375121/45522472-1d47bf80-b778-11e8-80dc-b0b21884cbcf.png">

1ec5 commented on platform/darwin/test/MGLSymbolStyleLayerTests.mm:
```diff
@@ -1096,20 +1098,20 @@ - (void)testProperties {
                               @"text should round-trip constant value expressions.");
 
-        constantExpression = [NSExpression expressionWithFormat:@"'Text Field'"];
+        constantExpression = [NSExpression expressionWithFormat:@"MGL_FUNCTION('format', 'Text Field', %@)", @{}];
```
Is it still possible to set the `textField` property to a string as opposed to a formatted string? If so, we should continue to test that capability alongside the tests for formatted strings. It might require some refactoring of the EJS code that generates this file.

ChrisLoer replied:
I believe that's what the first `layer.text = constantExpression` on line 1100 does? (see ` NSExpression *constantExpression = [NSExpression expressionWithFormat:@"'Text Field'"];`)

1ec5 replied:
The workflow I’m concerned about is existing code like this:

```objc
layer.text = [NSExpression expressionForConstantValue:@"Foo"];
// later
if ([layer.text.constantValue isEqualToString:@"Foo"]) { … } // crash
// or
if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Foo"]]) { … } // false
```

One always takes a risk by directly accessing `NSExpression.constantValue` without first checking the expression type, but if the developer set the `text` property in the first place, they may not expect to have to introduce that check when upgrading to a minor release of the SDK.

In other words, this implicit coercion is unusual in that it’s new. It’s also unusual in that formatted strings are a fairly foreign concept to iOS/macOS development, since we didn’t go the route of using NSAttributedString to represent them. Whereas coercions between strings and numbers are understood as invocations of the `CAST()` operator, there’s no `CAST()` operator support for converting between NSString and whatever class we’d represent a formatted string with. Introducing that `CAST()` support would be an elegant way of at least explaining away the surprise that some developers might experience.

Even so, we know from the v4.0.0 release that any changes to how coercions work can catch developers by surprise. In that case, we’re still fielding questions from developers who don’t expect to have to introduce `CAST()` into their inequality predicates. In this case, on the other hand, developers would be surprised at suddenly having to downcast from formatted strings to strings.

So my suggestion would be to automatically downcast from formatted strings to strings when there’s nothing but a `["formatted", "foo", {}]`. Extra `CAST()` support would be cool too.

ChrisLoer replied:
Right now the getter for _any_ constant expression does the downcast, so the cases you provided all return `true`. This is a little different from doing an automatic downcast based on whether there are options. Even if you add an explicit format expression, you'll still always get a string constant value back. Basically, `format` is almost invisible from the point of view of the iOS SDK with this PR, except that you can set it via `MGL_FUNCTION`.

Test code I used:
```
    MGLSymbolStyleLayer *layer = (MGLSymbolStyleLayer*)[self.mapView.style layerWithIdentifier:@"country-label"];
    layer.text = [NSExpression expressionForConstantValue:@"Foo"];
    // later
    if ([layer.text.constantValue isEqualToString:@"Foo"]) {
        NSLog(@"constantValue comparison worked");
    } // crash
    // or
    if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Foo"]]) {
        NSLog(@"constant expression comparison worked");
    }
    
    layer.text = [NSExpression expressionWithFormat:@"MGL_FUNCTION('format', 'Bar', %@)", @{}];
    // later
    if ([layer.text.constantValue isEqualToString:@"Bar"]) {
        NSLog(@"constantValue bar comparison worked");
    } // crash
    // or
    if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Bar"]]) {
        NSLog(@"constant bar expression comparison worked");
    }
```

ChrisLoer commented:
OK, I _think_ this is ready for review!

@1ec5, could you look at the darwin bindings? This is not meant to be a full set of bindings -- just enough to let existing usage keep working and enable use of `format` via `MGL_FUNCTION('format',...)`

@LukasPaczos, could you look at the Android bindings? I just hacked in an undocumented `Expression.format` method to allow the tests to pass, and I don't understand the tests well enough to be sure that all the round-tripping is still working. The idea is that when you set text-field, it will convert that to a `FormatExpression` under the hood, but when you get results out they'll convert back to plain strings for now.

@kkaefer Can I interest you in reviewing the core part? I'm not really sure who's best qualified to dig into all the expression serialization/conversion code (which I'm most iffy about). The actual shaping code @ansis might be well placed to review, but that part's not _too_ complicated and it's a pretty direct port from JS.

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/style/layer.junit.ejs:
```diff
@@ -165,7 +165,12 @@ public class <%- camelize(type) %>LayerTest extends BaseActivityTest {
 <% if (property.tokens) { -%>
 
       layer.setProperties(<%- camelizeWithLeadingLowercase(property.name) %>("{token}"));
+<% if (property.type === 'formatted') { -%>
```
I'd like to re-iterate concerns from https://github.com/mapbox/mapbox-gl-native/pull/12624#pullrequestreview-158773391 for Android as well. Will plain text, or other than `format` expressions accepted for `text-field` property? If yes, we need to rework this `.ejs` file to generate tests for both.

ChrisLoer replied:
Yeah plain text is accepted but may be transformed... 🤔

LukasPaczos replied:
Transformed to an expression? This would change the behavior of `PropertyValue#isExpression` when the `text-field` property is retrieved by `SymbolLayer#getTextField` which would return `false` even though it should return `true` for plain text. That's maybe a minor inconsistency but worth taking into consideration.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/expressions/Expression.java:
```diff
@@ -3224,6 +3224,11 @@ public static Expression collator(Expression caseSensitive, Expression diacritic
     return new Expression("collator", new ExpressionMap(map));
   }
 
+  public static Expression format(Expression input) {
```
If the plan is to introduce the full `format` expression implementation in a follow-up PR, I'd rather skip the public API + tests part here entirely.

ChrisLoer replied:
Hmm, like instead of modifying the test to use this, just omit the test entirely for the `text-field` case?

LukasPaczos replied:
Exactly, and remove the public API entry from `Expression` class.

ChrisLoer commented on platform/default/bidi.cpp:
```diff
@@ -100,6 +100,108 @@ std::vector<std::u16string> BiDi::processText(const std::u16string& input,
     return applyLineBreaking(lineBreakPoints);
 }
+    
+std::vector<StyledText> BiDi::processStyledText(const StyledText& input, std::set<std::size_t> lineBreakPoints) {
```
It would be pretty straightforward to add unit tests that were basically a native port of https://github.com/mapbox/mapbox-gl-rtl-text/blob/master/test/arabic.test.js, but this functionality is already exercised pretty directly by the `formatted/arabic` render test. When we can test effectively either way, I generally prefer render tests because they introduce less friction... what do you think?

ChrisLoer replied:
Added the test -- and it turned something up that render tests couldn't see! Lines that had characters removed from them during BiDi due to UBIDI_REMOVE_BIDI_CONTROLS could end up with extra null characters at the trailing end of the string (I don't think this would actually affect rendering, but it wasn't intentional and it'd be a surprise for anyone consuming these strings). I changed `BiDi::getLine` to match the behavior of `BiDi::writeReverse` and explicitly resize the output string before returning it.

ChrisLoer replied:
Even more interesting: the sanitizer found a problem in the ICU code?

`../../../vendor/icu/src/ubidiln.cpp:668:17: runtime error: left shift of 2 by 31 places cannot be represented in type 'int32_t' (aka 'int')`

ChrisLoer replied:
The sanitizer issue looks to me like it's legitimately undefined behavior in the C++ spec, although I have no idea if there are any compilers that would fail to get the last bit out the way the two problematic macros are trying to do. I looked on the ICU issue tracker and couldn't find an open issue, we should probably file one?

I tried applying a patch in our vendoring code, based on a patch used by LibreOffice: https://github.com/LibreOffice/core/blob/master/external/icu/icu4c-ubsan.patch.1 This is pretty brittle -- @kkaefer do you think it's reasonable until there's an upstream fix?

kkaefer replied:
Yeah, this fix looks good; let's apply it until there's an upstream fix.

ChrisLoer commented on src/mbgl/style/expression/value.cpp:
```diff
@@ -136,6 +143,31 @@ mbgl::Value ValueConverter<mbgl::Value>::fromExpressionValue(const Value& value)
             return mbgl::Value();
         },
+        [&](const Formatted& formatted)->mbgl::Value {
+            // TODO format: figure out right way to serialize this?
```
😳I missed that I had left these in! There are actually three linked TODOs and I could use some help thinking them through.

- `stringify(Writer& writer, const expression::Formatted& v)` writes a `Formatted` literal as the expression that would generate that literal, since there's no other way to represent a `Formatted` object.
- `value.cpp`'s `writeJSON` just calls through to the implementation in `stringify.hpp`, which I think makes sense, but I'm not sure if there's some context I'm missing about how that's supposed to work, because it seems like most of `writeJSON` is duplicating code in `stringify.hpp`
- The `(const Formatted& formatted)->mbgl::Value` converter does the same thing as the `stringify` implementation, except it's in the `mbgl::Value` format instead of writing out to a JSON writer. This seems like bad duplication. I'm somewhat following the pattern of `Color` (which has an `mbgl::Value` serialization and also a parallel `stringify`), but the duplication seems worse since the logic for `Formatted` is more complicated.

Maybe the right way to do this is to give `Formatted` a method that serializes it to an `mbgl::Value`, and then use that for all three situations (since there's already an `mbgl::Value` -> JSON serializer)?

ChrisLoer commented on src/mbgl/text/tagged_string.cpp:
```diff
@@ -0,0 +1,45 @@
+}
+    
+void TaggedString::verticalizePunctuation() {
+    // Relies on verticalization changing characters in place so that style indices don't need updating
```
It is currently true -- do you think it's unsafe to rely on?

https://github.com/mapbox/mapbox-gl-native/blob/d21c126f9def8627c2b33ec1eabde266cb9c42fe/src/mbgl/util/i18n.cpp#L562-L577

kkaefer replied:
I think it's fine to rely on, but we could add a comment to https://github.com/mapbox/mapbox-gl-native/blob/d21c126f9def8627c2b33ec1eabde266cb9c42fe/src/mbgl/util/i18n.cpp#L303-L321 indicating that we also need to change it here.

asheemmamoowala commented on src/mbgl/style/expression/format_expression.cpp:
```diff
@@ -0,0 +1,176 @@
+            }
+            FontStack fontStack;
+            for (const auto& value : textFontResult->get<mapbox::util::recursive_wrapper<std::vector<Value>>>().get()) {
+                // TODO format: is there an easier way to evaluate a fontStack?
```
nit: remove TODO

asheemmamoowala commented on src/mbgl/text/tagged_string.hpp:
```diff
@@ -0,0 +1,83 @@
+    FontStackHash fontStackHash;
+};
+
+struct TaggedString {
```
Comment would be useful to indicate that the `SectionOptions` and `TaggedString` structs represent evaluated values.

asheemmamoowala commented on include/mbgl/style/expression/coercion.hpp:
```diff
@@ -23,6 +23,8 @@ class Coercion : public Expression  {
 
     EvaluationResult evaluate(const EvaluationContext& params) const override;
     void eachChild(const std::function<void(const Expression&)>& visit) const override;
+    
```
The `Coercion` class-level comment at L14 is missing information about coercion of simple text to formatted text expressions.

ChrisLoer replied:
🤔 Actually I think that header comment may be outdated? It implies `Coercion` is just for `to-number` and `to-color`, which is no longer true. But it does make me wonder whether the default `Coercion` behavior of serial evaluation should really be supported for the `format` coercion -- since I don't think there's even any way to use it (the only way to create this coercion is through the implicit coerce on the `text-field` property).

ChrisLoer replied:
I moved the comment and disabled the coercion for `type::Formatted` (although this is a divergence from gl-js, there shouldn't actually be a way to exercise this code path either way, and this seems like a clearer expression of intent).

asheemmamoowala commented on src/mbgl/style/expression/format_expression.cpp:
```diff
@@ -0,0 +1,176 @@
+        const optional<Convertible> fontScaleOption = objectMember(options, "font-scale");
+        ParseResult fontScale;
+        if (fontScaleOption) {
+            fontScale = ctx.parse(*fontScaleOption, 1, {type::Number});
```
Will errors in parsing the font options return diagnosable errors ? It looks to me that the error would not mention `font-scale` or `text-font` when bailing early here?

ChrisLoer replied:
I _think_ this is following the pattern we use elsewhere -- the failure to parse `font-scale` or `text-font` will itself put a reportable error into the parsing context, and while that error won't say anything about `font-scale`, it will have that number-indexing way of telling what part of the expression failed. That number-indexing thing is pretty hard to read -- I usually spend a while guessing at the problem before I start trying to figure out what those numbers mean. But I think the right place to address that would be in the parsing/error code itself?

asheemmamoowala replied:
> that number-indexing way of telling what part of the expression failed 

Do you mean the indexing like this: `layers[14].filter[2][1]: string expected, array found`
If so I do think it should be improved in general.

ChrisLoer replied:
Yup that's what I meant.

asheemmamoowala commented on src/mbgl/text/shaping.cpp:
```diff
@@ -111,17 +115,25 @@ void justifyLine(std::vector<PositionedGlyph>& positionedGlyphs,
-        if (it != glyphs.end() && it->second) {
-            totalWidth += (*it->second)->metrics.advance + spacing;
+    for (std::size_t i = 0; i < logicalInput.length(); i++) {
+        const SectionOptions& section = logicalInput.getSection(i);
```
nit: `getSection` returns `SectionOptions`. I find the method names of `TaggedString` a bit hard to follow.

ChrisLoer replied:
Yeah, I see your point. Might make more sense to call it something like `getFormattingOptions`? At this point I'm following the JS naming conventions, but if we decide on better names here I can propagate them back to JS.

ChrisLoer commented:
> I couldn't figure out how the scaling is applied to the glyphs when passing to the shaders, but thats just me being curious.

The actual scaling is done in `quads.cpp` where the four corners of the glyph are scaled based on `positionedGlyph.scale`. This is different from what you might expect based on how we do data-driven `text-size` (looking at `a_size` in the shaders). The advantage of applying the scaling directly to the vertices is that we're not forced to create extra "size" buffers for every single symbol. The disadvantage is that we're not technically doing the same thing as `text-size` (e.g. if you use a really large `font-scale`, you'll notice that the halos are way too big). There's some discussion of this decision over in the JS PR.

ChrisLoer commented:
Sounds like two votes for `BiDi` unit tests -- I'll implement that tomorrow.

asheemmamoowala commented on src/mbgl/text/bidi.hpp:
```diff
@@ -14,17 +14,28 @@ class BiDiImpl;
+// the styling options to use for rendering that code point
+// The data structure is intended to accomodate the reordering/interleaving
+// of formatting that can happen when BiDi rearranges inputs
+using StyledText = std::pair<std::u16string, std::vector<uint8_t>>;
```
@ChrisLoer Since all symbol labels will be coerced to formatted text, are you concerned about memory consumption per-label increasing with each label being represented by a StyledText? The worst-case scenario is for a string to have a different style per-character, but I imagine that for the bulk of strings (those coerced) there will only be a single style. 
Would it be possible to maintain the styling options in a block format instead of per-code-point? the size of the vector would equal the number of sections and each entry would be the end index of the section with the corresponding index in the `sections` array. 

| Text              | hello | hola | こんにちは |
|-------------------|-------|------|------------|
| Section end index | 4     | 8    | 13?        |
| Section           | 0     | 1    | 2          |

`TaggedString` would need to perform a search linear in sections length for `getSectionAt()` but for the most common case that shouldn't be too bad.

ChrisLoer replied:
On the JS side I did a little experimentation with a data structure _kind of_ along the lines of what you're suggesting. It is a more memory-compact representation for the common case. What I was doing was iterating over the blocks as a whole, which made a lot of the iteration logic in the shaping code significantly more complicated. I think you're suggesting maintaining the same interface to `TaggedString` but just calculating the section index on-the-fly, so a CPU vs. memory trade off. That seems reasonable, but I don't know which would win without actually doing the experiment.

FWIW, on the JS side I looked at overall memory consumption in a full-screen map before and after this change, and couldn't detect any difference. It should be 1 byte-per-character, and I'd guess a full screen map has maybe on the order of 10k characters worth of text?

ChrisLoer commented:
OK, I think I've addressed all of the current round of review comments -- there were a lot of them, please let me know if I've left yours unresolved!

@asheemmamoowala I tried doing some memory profiling, but it's pretty hard to get reproducible numbers. There's at least no huge change. I think we just have to rely on the back-of-the-envelope estimate that says it shouldn't be too bad. The thing I worry about more in this PR is that it will lengthen tile layout/preparation time because of the extra format-related work. The good news is foregound/rendering costs should be pretty much unaffected since we're not adding any new GPU data or any extra foreground CPU processing.

@1ec5 Besides the existing tests, I played around a little with the `text` setter/getter to see it behaved the way I expected, and I tested that the macosapp's automatic localizer functionality still works without requiring any changes (testing that actually exposed a bug in my initial implementation of the aftermarket expression logic for `format`).

@LukasPaczos I went back to the "no Android wrapper for now" approach, with the assumption that we'll want to re-visit that when it's time to do a semver-major release. The constant-value getter/setters are still `String`-typed. The getter will now return an expression with the implicit `format` built-in. Instead of removing the tests for `text-field`, I added manual expression building code to the tests to handle `format` (e.g. `Converter.convert(new JsonArray().add("format").add(<rest of expression>))`)

1ec5 commented:
> This is not meant to be a full set of bindings -- just enough to let existing usage keep working and enable use of `format` via `MGL_FUNCTION('format',...)`

Do these changes account for the exception when using `MGL_FUNCTION` within an NSExpression or NSPredicate format string (#13026)?

ChrisLoer commented:
> Do these changes account for the exception when using MGL_FUNCTION within an NSExpression or NSPredicate format string (#13026)?

I've looked at that issue, but I don't understand how to trigger the problem (I'm sure you said how to do it, I just don't understand the vocabulary around NSExpression). FWIW, in terms of that issue, I'd expect the behavior here to be the same as the behavior for `MGL_FUNCTION('collator' ...` since I followed that pattern.

fabian-guerra commented:
@ChrisLoer this is an example of what makes `MGL_FUNCTION` crash:

```
NSExpression *gradientExpression = [NSExpression expressionWithFormat:@"mgl_interpolate:withCurveType:parameters:stops:(MGL_Function('line-progress'), 'linear', nil, %@)", stops];
```

kkaefer commented on include/mbgl/style/conversion/property_value.hpp:
```diff
@@ -28,6 +28,13 @@ struct Converter<PropertyValue<T>> {
     }
+    
+    PropertyValue<T> maybeConvertTokens(const expression::Formatted& t) const {
+        std::string todoText = t.sections[0].text;
```
This should be a `const std::string&` to avoid the copy.

kkaefer commented on scripts/vendor/icu.sh:
```diff
@@ -49,6 +49,12 @@ for FILE in "${ALL[@]}"; do
+sed -i.bak 's/((index)|((int32_t)(level)<<31))/((index)|((uint32_t)(level)<<31))/' "src/ubidiimp.h"
+sed -i.bak 's/((x)|=((int32_t)(level)<<31)/((x)|=((uint32_t)(level)<<31)/' "src/ubidiimp.h"
+rm -rf src/ubidiimp.h.bak
+
```
While this works, I think that using an actual patch conveys the intention of the command better. Additionally, we'll be alerted if the upstream file changes in an incompatible way when we're updating the ICU version.

```bash
patch -p0 << PATCH
--- src/ubidiimp.h
+++ src/ubidiimp.h
@@ -198,8 +198,8 @@
 /* in a Run, logicalStart will get this bit set if the run level is odd */
 #define INDEX_ODD_BIT (1UL<<31)
 
-#define MAKE_INDEX_ODD_PAIR(index, level) ((index)|((int32_t)(level)<<31))
+#define MAKE_INDEX_ODD_PAIR(index, level) ((index)|((uint32_t)(level)<<31))
-#define ADD_ODD_BIT_FROM_LEVEL(x, level)  ((x)|=((int32_t)(level)<<31))
+#define ADD_ODD_BIT_FROM_LEVEL(x, level)  ((x)|=((uint32_t)(level)<<31))
 #define REMOVE_ODD_BIT(x)                 ((x)&=~INDEX_ODD_BIT)
 
 #define GET_INDEX(x)   ((x)&~INDEX_ODD_BIT)
PATCH
```

ChrisLoer replied:
Ah yes, definitely better!

kkaefer commented on src/mbgl/text/glyph.hpp:
```diff
@@ -55,17 +55,21 @@ class Glyph {
 
 using Glyphs = std::map<GlyphID, optional<Immutable<Glyph>>>;
-using GlyphMap = std::map<FontStack, Glyphs>;
+using GlyphMap = std::map<FontStackHash, Glyphs>;
```
Curious, why change this to a hash instead of using `std::unordered_map` with a custom `Hash` template argument?

ChrisLoer replied:
It's true, the explicit hashing does feel awkward. I think the part that feels worst is when we have to do the reverse from hash->[string] for dependency lookup (i.e. `GeometryTileWorker::onGlyphsAvailable`).

I'm interested in the "store an iterator" idea, but I'm concerned that it will make the code harder to audit because you have to make sure that if you store an iterator you're using the same map when you try to use the iterator later -- and GeometryTileWorker already merges GlyphMaps from multiple requests to GlyphManager over time...
